### ISSUE TITLE
Strange caching behaviour with retry

### ISSUE THREAD

**Issue by TomiTakussaari** [sindresorhus/got#597]
Following test:

```javascript
/**
 * @jest-environment node
 */
import nock from 'nock';
import got from 'got';
import * as https from "https";
import * as http from "http";
import QuickLru from 'quick-lru';

const agentOpts = {
  keepAlive: true
};

const agents = {
  https: new https.Agent(agentOpts),
  http: new http.Agent(agentOpts)
};

const cache = new QuickLru({maxSize: 50});

const retry = {
  retries: 2,
  methods: ['GET'],
  statusCodes: [408, 413, 429, 502, 503, 504]
};

const requestConfig = {
  json: true,
  baseUrl: 'http://foobar/',
  timeout: 5000,
  agent: agents,
  cache: cache,
  retry
};

it('retries on http 502 error', async () => {
  const mocked = nock('http://foobar/')
    .get('/failure')
    .reply(502, JSON.stringify({success: false}))
    .get('/failure')
    .reply(200, JSON.stringify({success: true}));

  const response = await got('/failure', requestConfig);
  expect(mocked.isDone()).toEqual(true);
  console.log(response.fromCache);
  console.log(response.statusCode);
  expect(response.body).toEqual({success: true});
});
```
Fails with:
```
  console.log content/got.test.ts:45
    true

  console.log content/got.test.ts:46
    200


Error: expect(received).toEqual(expected)

Expected value to equal:
  {"success": true}
Received:
  {"success": false}
```

Which looks to me like Got is caching http-502 response, and then using it when it gets http-200 response.

If I change it like this:  

```javascript
.reply(502, JSON.stringify({success: false}), {`Cache-Control': 'no-store'})
```
Or disable caching completely, test passes fine.

This is not something I expected, and could be problematic.

First, I don't think that 502 responses should ever be cached: 

"Responses with status codes that are defined as cacheable by default (e.g., 200, 203, 204, 206, 300, 301, 404, 405, 410, 414, and 501 in this specification) can be reused by a cache with heuristic expiration unless otherwise indicated by the method definition or explicit cache controls [RFC7234]; all other status codes are not cacheable by default"

https://httpwg.org/specs/rfc7231.html#overview.of.status.codes

Secondly, why would cache even be used when retrying request ?

**Comment by szmarczak** [comment]
> I don't think that 502 responses should ever be cached

Got depends on the [`cacheable-request`](https://github.com/lukechilds/cacheable-request) module.

cc @lukechilds

**Comment by sindresorhus** [comment]
> Secondly, why would cache even be used when retrying request?

If this is correct, it sounds like a valid bug. It should not use cache for retrying.

**Comment by sindresorhus** [comment]
There's no 502 status code here:

> https://github.com/kornelski/http-cache-semantics/blob/d0bf1baf301dc6044a839ac309685e9385fb6c3b/index.js#L3

^ That the lib `cacheable-request` uses to determine cacheability.

**Comment by szmarczak** [comment]
> If this is correct, it sounds like a valid bug. It should not use cache for retrying.

It does :P I'll send a PR.

**Comment by szmarczak** [comment]
Fix: https://github.com/lukechilds/cacheable-request/pull/55

**Comment by lukechilds** [comment]
Sorry, autoclosed due to merging https://github.com/lukechilds/cacheable-request/pull/55

**Comment by lukechilds** [comment]
@szmarczak Released your fix in `cacheable-request@5.0.0`.

**Comment by TomiTakussaari** [comment]
Thanks! Seems to be working now.
